### PR TITLE
Add SmolLM (smollm2)

### PR DESCRIPTION
**Pull request by Inklingdq**
### Summary
Add SmolLM 135M model (smollm2) for issue https://github.com/pytorch/executorch/issues/9324

```
python -m examples.models.llama.runner.native --model smollm2 \
--pte smollm2.pte  \
--tokenizer /Users/danqingwang/tmp/snapshots/1d461723eec654e65efdc40cf49301c89c0c92f4/tokenizer.json \
--tokenizer_config /Users/danqingwang/tmp/snapshots/1d461723eec654e65efdc40cf49301c89c0c92f4/tokenizer_config.json \
--prompt "What ingredients are in a California roll?" \
--params examples/models/smollm2/135M_config.json --max_len 64 \
--temperature 0 -kv
import error: No module named 'triton'
W0318 22:08:14.883045 53671 site-packages/torch/distributed/elastic/multiprocessing/redirects.py:29] NOTE: Redirects are currently not supported in Windows or MacOs.
[program.cpp:136] InternalConsistency verification requested but not available

California rolls are made with a variety of ingredients, including wheat, corn, rice, and beans. The ingredients are combined in a special way to create a unique texture and flavor.
What is the difference between a California roll and a roll?
The main difference between a

Prefill time: 0.03980898857116699
Generation tok/s: 76.69897827254086
```
output agrees with the eager model output 
image

### Test plan
Convert to meta format
```
python examples/models/smollm/convert_weights.py /Users/danqingwang/tmp/snapshots/1d461723eec654e65efdc40cf49301c89c0c92f4/ /Users/danqingwang/smollm.pth
```

Run export 
```
./install_executorch.sh --pybind xnnpack && python -m examples.models.llama.export_llama   \
--model smollm2 --params examples/models/smollm2/135M_config.json  \
--checkpoint /Users/danqingwang/smollm.pth -kv --use_sdpa_with_kv_cache  \
-X -d fp32 --metadata '{"get_bos_id":[11191, 12870], "get_eos_ids":[29070, 25853]}'  \
--output_name="smollm2.pte" --verbose
```

Run test
python -m examples.models.llama.runner.native --model smollm2 \
--pte smollm2.pte  \
--tokenizer /Users/danqingwang/tmp/snapshots/1d461723eec654e65efdc40cf49301c89c0c92f4/tokenizer.json \
--tokenizer_config /Users/danqingwang/tmp/snapshots/1d461723eec654e65efdc40cf49301c89c0c92f4/tokenizer_config.json \
--prompt "What ingredients are in a California roll?" \
--params examples/models/smollm2/135M_config.json --max_len 64 \
--temperature 0 -kv

### CODE REVIEW

File: examples/models/smollm/135M_config.json [hunk 1]
```diff
@@ -0,0 +1,14 @@
+{
+    "dim": 576,
+    "ffn_dim_multiplier": 1,
```
Comment (Inklingdq) [inline examples/models/smollm/135M_config.json:3]: There's some size mismatch error during quantization
```
	size mismatch for layers.0.feed_forward.w1.weight: copying a param with shape torch.Size([1536, 576]) from checkpoint, the shape in current model is torch.Size([576, 576]).
	size mismatch for layers.0.feed_forward.w2.weight: copying a param with shape torch.Size([576, 1536]) from checkpoint, the shape in current model is torch.Size([576, 576]).
	size mismatch for layers.0.feed_forward.w3.weight: copying a param with shape torch.Size([1536, 576]) from checkpoint, the shape in current model is torch.Size([576, 576]).
```
I'm not very sure about the definiation of dim and ffn_dim_multiplier here, looks like some wrong value here? Would you mind provide some pointer/context on this? Appreciate it! @jackzhxng 

The model structure is below

LlamaForCausalLM(
  (model): LlamaModel(
    (embed_tokens): Embedding(49152, 576)
    (layers): ModuleList(
      (0-29): 30 x LlamaDecoderLayer(
        (self_attn): LlamaSdpaAttention(
          (q_proj): Linear(in_features=576, out_features=576, bias=False)
          (k_proj): Linear(in_features=576, out_features=192, bias=False)
          (v_proj): Linear(in_features=576, out_features=192, bias=False)
          (o_proj): Linear(in_features=576, out_features=576, bias=False)
          (rotary_emb): LlamaRotaryEmbedding()
        )
        (mlp): LlamaMLP(
          (gate_proj): Linear(in_features=576, out_features=1536, bias=False)
          (up_proj): Linear(in_features=576, out_features=1536, bias=False)
          (down_proj): Linear(in_features=1536, out_features=576, bias=False)
          (act_fn): SiLU()
        )
        (input_layernorm): LlamaRMSNorm((576,), eps=1e-05)
        (post_attention_layernorm): LlamaRMSNorm((576,), eps=1e-05)
      )
    )
    (norm): LlamaRMSNorm((576,), eps=1e-05)
    (rotary_emb): LlamaRotaryEmbedding()
  )
  (lm_head): Linear(in_features=576, out_features=49152, bias=False)
)

File: examples/models/smollm/convert_weights.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    parser.add_argument("output", type=str, help="Path to the output checkpoint")
+
+    args = parser.parse_args()
+
+    # Don't necessarily need to use TorchTune checkpointer, can just aggregate checkpoint files by ourselves.
+    checkpointer = FullModelHFCheckpointer(
+        checkpoint_dir=args.input_dir,
+        checkpoint_files=["model.safetensors"],
+        output_dir=".",
+        model_type="MISTRAL",
```
Comment (jackzhxng) [inline examples/models/smollm/convert_weights.py:71]: Change to Llama

File: examples/models/smollm/convert_weights.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        Dict[str, torch.Tensor]: State dict in Meta's format.
+    """
+    converted_state_dict = {}
+    inverted_mapping_dict = {v: k for k, v in _SMOLLM_FROM_META.items()}
+    for key, value in state_dict.items():
+        new_key = get_mapped_key(key, inverted_mapping_dict)
+        converted_state_dict[new_key] = value
+
+    # Input and output embeddings are tied.
+    converted_state_dict["output.weight"] = converted_state_dict[
```
Comment (jackzhxng) [inline examples/models/smollm/convert_weights.py:46]: Might be because of this, input and output embeddings are not shared for Llama which this model is based off of

File: examples/models/llama/export_llama_lib.py [hunk 4]
```diff
@@ -94,6 +94,7 @@
     "static_llama",
     "qwen2_5",
     "phi-4-mini",
+    "smollm",
```
Comment (jackzhxng) [inline examples/models/llama/export_llama_lib.py:97]: rename this and directory to smolllm2

Comment (Inklingdq) [reply examples/models/llama/export_llama_lib.py:97]: Thanks! Should it be smollm2 or smolllm2?

File: examples/models/smollm/convert_weights.py (hunk 2 above)
Comment (Inklingdq) [reply examples/models/smollm/convert_weights.py:71]: Thank you updated~

File: examples/models/smollm/convert_weights.py (hunk 3 above)
Comment (Inklingdq) [reply examples/models/smollm/convert_weights.py:46]: Make sense, removed this

File: examples/models/llama/export_llama_lib.py (hunk 4 above)
Comment (jackzhxng) [reply examples/models/llama/export_llama_lib.py]: Ah - it should be smollm2*

File: examples/models/smollm/135M_config.json [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+    "hidden_dim": 576,
```
Comment (jackzhxng) [inline examples/models/smollm/135M_config.json:4]: Should be 1536 - https://huggingface.co/HuggingFaceTB/SmolLM2-135M/blob/main/config.json#L12

Comment (Inklingdq) [reply examples/models/smollm/135M_config.json:4]: Thank you! I got it mixed with the hidden_size 😄

File: examples/models/smollm/135M_config.json [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+    "n_heads": 9,
+    "n_kv_heads": 3,
+    "n_layers": 30,
+    "norm_eps": 1e-05,
+    "rope_theta": 10000.0,
+    "use_scaled_rope": false,
+    "vocab_size": 49152,
+    "use_hf_rope": true,
```
Comment (jackzhxng) [inline examples/models/smollm/135M_config.json:12]: this should be `false`

Comment (Inklingdq) [reply examples/models/smollm/135M_config.json]: Thank you!! Updated

**Review by jackzhxng — APPROVED** [review]
Thanks for helping with this!

**Comment by Inklingdq** [comment]
hi @jackzhxng, it looks like the 3 failing checks are not related to my change, anyway to rerun the test?

**Comment by jackzhxng** [comment]
@larryliu0820 yeah I'm not sure how this got into viable/strict, I think I just didn't see it, but as you can see above it's been reverted and picked back into main already
